### PR TITLE
Allow generating source maps.

### PR DESCRIPTION
**Pull request by gdotdesign**
This PR builds on and supersedes #356 (Mainly the VLQ implementation is brought over).
- SPEC: https://tc39.es/source-map/
- Testing: https://evanw.github.io/source-map-visualization/ https://sokra.github.io/source-map-visualization/
- I've tested this with `mint-website` and `mint-ui-website` and it seems to be working.
- It's slow at this time (~1s from a ~2.2s build).

--------

Developer Notes:
- Because of the new compiler architecture it was relatively easy to implement.
- Changes in specs are because the order of resolving entities changed.

TODO:
- [x] Mappings and sources
- [x] Wiring to commands (`--generate-source-maps` flag)
- [x] Symbol names for the debugger 
- [x] Cleanup pass

The reason for the slowness is `Ast::Node#location` calculation from offset to line/column, which needs to be calculated when parsing so it will be in a separate PR.

### CODE REVIEW

File: src/utils/vlq.cr
Comment (Sija) [inline src/utils/vlq.cr:1]: Perhaps it would make sense to extract it as a separate shard?

Comment (gdotdesign) [reply src/utils/vlq.cr:1]: Ideally yes, but I'm not up to do it (and keep maintaining it).

File: src/utils/source_map_generator.cr [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    getter sources = [] of String
+
+    # The generate JavaScript file.
+    getter generated : String
+
+    def initialize(unsorted : Deque(Mapping), @generated)
+      # We sort the mappings into lines.
+      unsorted.each do |item|
+        mappings[item[1][0]] ||= [] of Mapping
+        mappings[item[1][0]] << item
```
Comment (Sija) [inline src/utils/source_map_generator.cr:25]:
```suggestion
        (mappings[item[1][0]] ||= [] of Mapping) << item
```

File: src/utils/source_map_generator.cr [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    def generate : String
+      last_name_index = 0
+      last_column = 0
+      last_index = 0
+      last_line = 0
+
+      generated_mappings =
+        (0..generated.lines.size).map do |line_index|
+          if items = mappings[line_index]?
```
Comment (Sija) [inline src/utils/source_map_generator.cr:52]:
```suggestion
          next "" unless items = mappings[line_index]?
```

File: src/utils/source_map_generator.cr [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                last_index = source_index
+                last_line = source_line
+              end
+
+              memo << segment
+              {memo, column}
+            end[0].join(",")
+          else
+            ""
+          end
```
Comment (Sija) [inline src/utils/source_map_generator.cr:98]:
```suggestion
```

File: src/utils/source_map_generator.cr [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                end
+
+                last_column = source_column
+                last_index = source_index
+                last_line = source_line
+              end
+
+              memo << segment
+              {memo, column}
+            end[0].join(",")
```
Comment (Sija) [inline src/utils/source_map_generator.cr:95]:
```suggestion
            end.first.join(",")
```

File: src/utils/vlq.cr [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+      while continue
+        char = chars.shift or raise "Expected more digits in base 64 VLQ value."
+        digit = base64_decode(char)
+        continue = false if (digit & VLQ_CONTINUATION_BIT) == 0
+        digit &= VLQ_BASE_MASK
+        vlq += digit << shift
+        shift += VLQ_BASE_SHIFT
+      end
+
+      [from_vlq_signed(vlq), String.new(chars)]
```
Comment (Sija) [inline src/utils/vlq.cr:70]:
```suggestion
      {from_vlq_signed(vlq), String.new(chars)}
```

File: src/utils/vlq.cr [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+    # Decode an array of variable length quantities from the given string
+    def self.decode_array(str)
+      output = [] of Tuple(Int32, String)
+      while !str.empty?
+        int, str = decode(str)
+        output << int
+      end
+      output
+    end
```
Comment (Sija) [inline src/utils/vlq.cr:81]: It seems you're creating an array of tuples, yet you append ints to it?

Comment (gdotdesign) [reply src/utils/vlq.cr]: I didn't check this class, just brought it over. We don't decode anything, so this code was not even checked.

I removed the decoding part.

**Review by Sija — APPROVED** [review]

**Pull request merged**